daemons/gptp/common/ieee1588clock.cpp

```diff
@@ -350,26 +358,44 @@ void IEEE1588Clock::setMasterOffset
 	}
 
 	if( _syntonize ) {
-		if( _new_syntonization_set_point ) {
+		if( _new_syntonization_set_point || _phase_error_violation > PHASE_ERROR_MAX_COUNT ) {
 			_new_syntonization_set_point = false;
+			_phase_error_violation = 0;
 			if( _timestamper ) {
 				/* Make sure that there are no transmit operations
 				   in progress */
 				getTxLockAll();
+				if (port->getTestMode()) {
+					GPTP_LOG_STATUS("Adjust clock phase offset:%lld", -master_local_offset);
+				}
 				_timestamper->HWTimestamper_adjclockphase
 					( -master_local_offset );
 				_master_local_freq_offset_init = false;
+				restartPDelayAll();
 				putTxLockAll();
 				master_local_offset = 0;
 			}
 		}
+
 		// Adjust for frequency offset
 		long double phase_error = (long double) -master_local_offset;
-		_ppm += (float) (INTEGRAL*phase_error +
-			 PROPORTIONAL*((master_local_freq_offset-1.0)*1000000));
+		if( fabsl(phase_error) > PHASE_ERROR_THRESHOLD ) {
+			++_phase_error_violation;
+		} else {
+			_phase_error_violation = 0;
+
+			float syncPerSec = (float)(1.0 / pow((float)2, port->getSyncInterval()));
```

Comment:
As far as I know this is the only usage of a sync per second. If it is being done in other places I would agree also that it could be moved to the port class.
There was previously some discussion about the use of pow() to resolve the base 2 intervals. This pattern is used in all occurrences in gPTP. For consistency I would favor to keep them all the same. That said if someone wanted to go and update the usage throughout gPTP to something else I would support that as long as it remains consistent. Along with that perhaps it would make sense if someone does do that to add interfaces to the port class to return either the base 2 value or interval in time (ms, us or ns).